arch/xtensa/core/mpu.c

```diff
@@ -0,0 +1,755 @@
+ *
+ * @param entries Array of MPU entries with XTENSA_MPU_NUM_ENTRIES elements.
+ *
+ * @return Index of the first enabled entry.
+ * @retval XTENSA_MPU_NUM_ENTRIES if no entry is enabled.
+ */
+static inline uint8_t find_first_enabled_entry(const struct xtensa_mpu_entry *entries)
+{
+	int first_enabled_idx;
+
+	for (first_enabled_idx = 0; first_enabled_idx < XTENSA_MPU_NUM_ENTRIES;
+	     first_enabled_idx++) {
+		if (entries[first_enabled_idx].as.p.enable) {
+			break;
+		}
+	}
+
+	return first_enabled_idx;
+}
+
+/**
+ * Consolidate the MPU entries.
+ *
+ * This removes consecutive entries where the attributes are the same.
+ *
+ * @param entries Array of MPU entries with XTENSA_MPU_NUM_ENTRIES elements.
+ * @param first_enabled_idx Index of first enabled entry.
+ *
+ * @return Index of the first enabled entry after consolidation.
+ */
+static uint8_t consolidate_entries(struct xtensa_mpu_entry *entries,
+				   uint8_t first_enabled_idx)
+{
+	uint8_t new_first = first_enabled_idx;
+	uint8_t idx = first_enabled_idx + 1;
+
+	/* For each a pair of entries... */
+	while (idx < XTENSA_MPU_NUM_ENTRIES) {
+		struct xtensa_mpu_entry *entry_0 = &entries[idx - 1];
+		struct xtensa_mpu_entry *entry_1 = &entries[idx];
+		bool remove = false;
+
+		if (xtensa_mpu_entries_has_same_attributes(entry_0, entry_1)) {
+			/*
+			 * If both entry has same attributes (access_rights and memory type),
+			 * they can be consolidated into one.
+			 */
+			remove = true;
+		} else if (xtensa_mpu_entries_has_same_address(entry_0, entry_1)) {
+			/*
+			 * If both entries have the same address, the higher index
+			 * one always override the lower one. So remove the lower index
+			 * one.
+			 */
+			remove = true;
+		}
+
+		if (remove) {
+			mpu_entries_shift_one_down(entries, new_first, idx);
+
+			new_first++;
+
+			continue;
+		}
+
+		idx++;
+	}
+
+	return new_first;
+}
+
+/**
+ * Add a memory region to the MPU map.
+ *
+ * This adds a memory region to the MPU map, by setting the appropriate
+ * start and end entries. This may re-use existing entries or add new
+ * entries to the map.
+ *
+ * @param[in,out] map Pointer to MPU map.
+ * @param[in] start_addr Start address of the region.
+ * @param[in] end_addr End address of the region.
+ * @param[in] access_rights Access rights of this region.
+ * @param[in] memory_type Memory type of this region.
+ * @param[out] first_idx Return index of first enabled entry if not NULL.
+ *                       If this is not NULL, and the value is the same as
+ *                       XTENSA_MPU_NUM_ENTRIES, this means this is
+ *                       an empty map. So the entry is added to map as-is.
+ *
+ * @retval 0 Successful in adding the region.
+ * @retval -EINVAL Invalid values in function arguments.
+ */
+static int mpu_map_region_add(struct xtensa_mpu_map *map,
+			      uintptr_t start_addr, uintptr_t end_addr,
+			      uint32_t access_rights, uint32_t memory_type,
+			      uint8_t *first_idx)
```

Comment:
This function still looks complicated, though I do see that it's no longer doing any sorting internally.  But it's still doing a **lot** of work trying to keep the mpu entries in a correctly formatted form at all times, every time you want to modify even a single one of them.  Seems like it would be simpler to just keep a list of regions to "compile" to a single array and do it all at once?  That would require a failure path in the case where you try to compile an invalid map or whatever.  But AFAICT userspace is tolerant of that.